packages/flutter_tools/lib/src/test/coverage_collector.dart

```diff
@@ -141,3 +153,136 @@ class CoverageCollector extends TestWatcher {
     return true;
   }
 }
+
+Future<Map<String, dynamic>> collect(Uri serviceUri, bool Function(String) libraryPredicate) async {
+  final VMService vmService = await VMService.connect(serviceUri, compression: CompressionOptions.compressionOff);
+  await vmService.getVM();
+  return _getAllCoverage(vmService, libraryPredicate);
+}
+
+
+Future<Map<String, dynamic>> _getAllCoverage(VMService service, bool Function(String) libraryPredicate) async {
+  await service.getVM();
+  final List<Map<String, dynamic>> coverage = <Map<String, dynamic>>[];
+  for (Isolate isolateRef in service.vm.isolates) {
+    await isolateRef.load();
+    final Map<String, dynamic> scriptList = await isolateRef.invokeRpcRaw('getScripts', params: <String, dynamic>{'isolateId': isolateRef.id});
+    final List<Future<void>> futures = <Future<void>>[];
+
+    final Map<String, Map<String, dynamic>> scripts = <String, Map<String, dynamic>>{};
+    final Map<String, Map<String, dynamic>> sourceReports = <String, Map<String, dynamic>>{};
+    // For each ScriptRef loaded into the VM, load the corresponding Script and
+    // SourceReport object.
+    for (Map<String, dynamic> script in scriptList['scripts']) {
+      if (!libraryPredicate(script['uri'])) {
+        continue;
+      }
+      final String scriptId = script['id'];
+      futures.add(
+        isolateRef.invokeRpcRaw('getSourceReport', params: <String, dynamic>{
+          'forceCompile': true,
+          'scriptId': scriptId,
+          'isolateId': isolateRef.id,
+          'reports': <String>['Coverage'],
+        })
+        .then((Map<String, dynamic> report) {
+          sourceReports[scriptId] = report;
+        })
+      );
+      futures.add(
+        isolateRef.invokeRpcRaw('getObject', params: <String, dynamic>{
+          'isolateId': isolateRef.id,
+          'objectId': scriptId,
+        })
+        .then((Map<String, dynamic> script) {
+          scripts[scriptId] = script;
+        })
+      );
+    }
+    await Future.wait(futures);
+    _buildCoverageMap(scripts, sourceReports, coverage);
+  }
+  return <String, dynamic>{'type': 'CodeCoverage', 'coverage': coverage};
+}
+
+// Build a hitmap of Uri -> Line -> Hit Count for each script object.
+void _buildCoverageMap(
+  Map<String, Map<String, dynamic>> scripts,
+  Map<String, Map<String, dynamic>> sourceReports,
+  List<Map<String, dynamic>> coverage,
+) {
+  final Map<Uri, Map<int, int>> hitMaps = <Uri, Map<int, int>>{};
+  for (String scriptId in scripts.keys) {
+    final Map<String, dynamic> sourceReport = sourceReports[scriptId];
+    for (Map<String, dynamic> range in sourceReport['ranges']) {
+      final Map<String, dynamic> coverage = range['coverage'];
+      final Map<String, dynamic> scriptRef = sourceReport['scripts'][range['scriptIndex']];
+      final Uri uri = Uri.parse(scriptRef['uri']);
```

Comment:
Do you have to convert scriptRef['uri'] to a Uri object? It looks like all that's done with it is converting it back to a string in _toScriptCoverageJson.